packages/injectable/ogre-tools-injectable.d.ts

```diff
@@ -1,160 +1,144 @@
 /// <reference types="jest" />
 declare module '@ogre-tools/injectable' {
-  type InferFromInjectable<T> = T extends NormalInjectable<
-    infer TInstance,
-    infer TInstantiationParameter
-  >
-    ? [TInstance, TInstantiationParameter]
-    : never;
-
-  type ValueType =
-    | string
-    | number
-    | boolean
-    | symbol
-    | bigint
-    | object
-    | Array<any>;
-
-  export type TentativeTuple<T> = T extends ValueType ? [T] : [undefined?];
-
   export interface DiContainer extends DiContainerForInjection<false> {
-    purge: (injectableKey: NormalInjectable<any, any>) => void;
+    purge: (injectableKey: Injectable<any, any, any>) => void;
 
     runSetups: () => Promise<void>;
 
-    override<TInjectable extends NormalInjectable<unknown, unknown>>(
-      injectable: TInjectable,
-      instantiateStub: (
-        di: DiContainerForInstantiate,
-        ...instantiationParameter: TentativeTuple<
-          InferFromInjectable<TInjectable>[1]
-        >
-      ) => InferFromInjectable<TInjectable>[0],
+    override<
+      Instance extends InjectionInstance,
+      InjectionInstance,
+      InstantiationParam
+    >(
+      injectable: Injectable<Instance, InjectionInstance, InstantiationParam>,
+      instantiateStub: Instantiate<Instance, InstantiationParam>,
     ): void;
 
-    register(injectable: NormalInjectable<any, any>): void;
-    register(injectable: InjectionTokenInjectable<any>): void;
+    register<
+      InjectionInstance extends Instance,
+      Instance,
+      InstantiationParam
+    >(injectable: Injectable<InjectionInstance, Instance, InstantiationParam>): void;
     preventSideEffects: () => void;
   }
 
-  export interface DiContainerForSetup extends DiContainerForInjection<true> {}
+  export type Instantiate<Instance, InstantiationParam> = {
+    (di: DiContainerForInstantiate, param: InstantiationParam extends void
+      ? void
+      : InstantiationParam
+    ): Instance;
+  };
 
-  export interface DiContainerForInstantiate
-    extends DiContainerForInjection<false> {}
+  export type DiContainerForSetup = DiContainerForInjection<true>;
+  export type DiContainerForInstantiate = DiContainerForInjection<false>;
 
-  type InjectionToken<TInstance, TInstantiationParameter> = {
-    template: TInstance;
-    instantiationParameter: TInstantiationParameter;
+  export interface InjectionToken<
+    Instance,
+    InstantiationParam = void,
+  > {
+    template: Instance;
+    instantiationParameter: InstantiationParam;
     key: Symbol;
-  };
+  }
+
+  export type OptionalParam<InstantiationParam> = InstantiationParam extends void
+    ? []
+    : [InstantiationParam];
```

Comment:
Naming convention: `InstantiationParameter` -> `TInstantiationParameter`. We'll consolidate this.